### PR TITLE
Add more span-based Vector<T> ctors and CopyTo methods

### PR DESCRIPTION
**Pull request by GrabYourPitchforks**
See https://github.com/dotnet/corefx/issues/30968 for full context.

Note to reviewers: I put type checks around `{Try}CopyTo(ROS<byte>)` because otherwise you could have a weird edge case where code like `default(Vector<char>).CopyTo(new Span<byte>(new byte[64]))` compiles and runs, but pretty much all other methods on `Vector<char>` fail.

These checks aren't needed around methods which reference the static `Count` property directly (using the implicit _T_ from the current _Vector\<T\>_) because the type initializer already fails in those cases and causes the expected runtime error. For example, `default(Vector<char>).CopyTo(new char[64])` - which uses the existing `CopyTo(T[])` method - isn't protected by an explicit check, but the `Count` property access near the top of the method is enough to trigger the expected runtime failure.

### CODE REVIEW

**Comment by GrabYourPitchforks** [comment]
Reference APIs and unit tests over at https://github.com/dotnet/corefx/pull/36136.

File: src/System.Private.CoreLib/shared/System/Numerics/Vector.cs [hunk 1]
```diff
@@ -778,7 +778,65 @@ private Vector(ref Register existingRegister)
...
+                || (typeof(T) == typeof(uint))
+                || (typeof(T) == typeof(int))
+                || (typeof(T) == typeof(ulong))
+                || (typeof(T) == typeof(long))
+                || (typeof(T) == typeof(float))
+                || (typeof(T) == typeof(double)))
+            {
+                if (values.Length < Vector<byte>.Count)
+                {
+                    throw new IndexOutOfRangeException(SR.Format(SR.Arg_InsufficientNumberOfElements, Vector<byte>.Count, nameof(values)));
```
Comment (omariom) [inline src/System.Private.CoreLib/shared/System/Numerics/Vector.cs]: :usual_suspects: Put throw statements to  static methods. It will make body of the method much smaller.

File: src/System.Private.CoreLib/shared/System/Numerics/Vector.cs [hunk 2]
```diff
@@ -778,7 +778,65 @@ private Vector(ref Register existingRegister)
 
 #if netcoreapp
         /// <summary>
-        /// Constructs a vector from the given span. The span must contain at least Vector'T.Count elements.
+        /// Constructs a vector from the given <see cref="ReadOnlySpan{Byte}"/>. The span must contain at least <see cref="Vector{Byte}.Count"/> elements.
+        /// </summary>
+        public Vector(ReadOnlySpan<byte> values)
+            : this()
+        {
+            if ((typeof(T) == typeof(byte))
```
Comment (tannergooding) [inline src/System.Private.CoreLib/shared/System/Numerics/Vector.cs]: nit: Would be nice to extract these into an `IsSupported` check like we did for `Vector128<T>` (and others):
https://source.dot.net/#System.Private.CoreLib/shared/System/Runtime/Intrinsics/Vector128_1.cs,76

File: src/System.Private.CoreLib/shared/System/Numerics/Vector.cs [hunk 3]
```diff
@@ -778,7 +778,65 @@ private Vector(ref Register existingRegister)
...
+            {
+                if (values.Length < Vector<byte>.Count)
+                {
+                    throw new IndexOutOfRangeException(SR.Format(SR.Arg_InsufficientNumberOfElements, Vector<byte>.Count, nameof(values)));
+                }
+                this = Unsafe.ReadUnaligned<Vector<T>>(ref MemoryMarshal.GetReference(values));
+            }
+            else
+            {
+                throw new NotSupportedException(SR.Arg_TypeNotSupported);
```
Comment (tannergooding) [inline src/System.Private.CoreLib/shared/System/Numerics/Vector.cs]: We have a `ThrowForUnsupportedVectorBaseType` that could be used: https://source.dot.net/#System.Private.CoreLib/shared/System/ThrowHelper.cs,407

We should probably add a note that it is also used for `Vector<T>` so it isn't expanded without consideration for Numerics and Intrinsics.

File: src/System.Private.CoreLib/shared/System/Numerics/Vector.cs (hunk 1 above)
Comment (tannergooding) [inline src/System.Private.CoreLib/shared/System/Numerics/Vector.cs]: Isn't the normal recommendation to throw `ArgumentOutOfRange` rather than `IndexOutOfRange`?

File: src/System.Private.CoreLib/shared/System/Numerics/Vector.cs [hunk 4]
```diff
@@ -809,6 +867,52 @@ public Vector(Span<T> values)
         #endregion Constructors
 
         #region Public Instance Methods
+        /// <summary>
+        /// Copies the vector to the given <see cref="Span{Byte}"/>. The destination span must be at least size <see cref="Vector{Byte}.Count"/>.
+        /// </summary>
+        /// <param name="destination">The destination span which the values are copied into</param>
+        /// <exception cref="ArgumentException">If number of elements in source vector is greater than those available in destination span</exception>
+        public void CopyTo(Span<byte> destination)
```
Comment (tannergooding) [inline src/System.Private.CoreLib/shared/System/Numerics/Vector.cs]: What does the codegen look like for these methods currently?

File: src/System.Private.CoreLib/shared/System/Numerics/Vector.cs (hunk 1 above)
Comment (GrabYourPitchforks) [reply src/System.Private.CoreLib/shared/System/Numerics/Vector.cs]: Probably. I didn't feel this was the right place to deviate from the established pattern, though. Maybe a good candidate for when we do a sweep of the .tt file later?

File: src/System.Private.CoreLib/shared/System/Numerics/Vector.cs (hunk 2 above)
Comment (GrabYourPitchforks) [reply src/System.Private.CoreLib/shared/System/Numerics/Vector.cs]: Do you know if this would mess with the codegen? I experimented and didn't see a difference in my limited testing, but I also don't know the full breadth of scenarios where codegen might produce suboptimal results.

Comment (tannergooding) [reply src/System.Private.CoreLib/shared/System/Numerics/Vector.cs]: We are already doing it in the HWIntrinsics, so I don't believe it does.

File: src/System.Private.CoreLib/shared/System/Numerics/Vector.cs (hunk 1 above)
Comment (tannergooding) [reply src/System.Private.CoreLib/shared/System/Numerics/Vector.cs]: Sounds good.

File: src/System.Private.CoreLib/shared/System/Numerics/Vector.cs (hunk 4 above)
Comment (GrabYourPitchforks) [reply src/System.Private.CoreLib/shared/System/Numerics/Vector.cs]: I'm working on getting the disasm and will post it in the main comments.

**Comment by GrabYourPitchforks** [comment]
Test method and codegen:

```cs
[MethodImpl(MethodImplOptions.NoInlining)]
private static uint SumAndGetFirstElement(ReadOnlySpan<byte> bytes, ReadOnlySpan<uint> uints)
{
    return (new Vector<uint>(bytes) + new Vector<uint>(uints))[0];
}
```

```asm
ConsoleAppBenchmark!ConsoleAppBenchmark.VectorRunner.SumAndGetFirstElement(System.ReadOnlySpan`1<Byte>, System.ReadOnlySpan`1<UInt32>):
00007ffc`437f0770 4883ec28        sub     rsp,28h
00007ffc`437f0774 c5f877          vzeroupper
00007ffc`437f0777 488b01          mov     rax,qword ptr [rcx]
00007ffc`437f077a 8b4908          mov     ecx,dword ptr [rcx+8]
00007ffc`437f077d 83f920          cmp     ecx,20h
00007ffc`437f0780 7c23            jl      ConsoleAppBenchmark!ConsoleAppBenchmark.VectorRunner.SumAndGetFirstElement(System.ReadOnlySpan`1<Byte>, System.ReadOnlySpan`1<UInt32>)+0x35 (00007ffc`437f07a5)
00007ffc`437f0782 c5fd1000        vmovupd ymm0,ymmword ptr [rax]
00007ffc`437f0786 488b02          mov     rax,qword ptr [rdx]
00007ffc`437f0789 8b4a08          mov     ecx,dword ptr [rdx+8]
00007ffc`437f078c 83f908          cmp     ecx,8
00007ffc`437f078f 7c1f            jl      ConsoleAppBenchmark!ConsoleAppBenchmark.VectorRunner.SumAndGetFirstElement(System.ReadOnlySpan`1<Byte>, System.ReadOnlySpan`1<UInt32>)+0x40 (00007ffc`437f07b0)
00007ffc`437f0791 c5fd1008        vmovupd ymm1,ymmword ptr [rax]
00007ffc`437f0795 c5fdfec1        vpaddd  ymm0,ymm0,ymm1
00007ffc`437f0799 c5f97ec0        vmovd   eax,xmm0
00007ffc`437f079d c5f877          vzeroupper
00007ffc`437f07a0 4883c428        add     rsp,28h
00007ffc`437f07a4 c3              ret
00007ffc`437f07a5 b920000000      mov     ecx,20h
00007ffc`437f07aa e869a4d8ff      call    System.Numerics.Vector.ThrowInsufficientNumberOfElementsException(Int32) (00007ffc`4357ac18)
00007ffc`437f07af cc              int     3
00007ffc`437f07b0 b908000000      mov     ecx,8
00007ffc`437f07b5 e85ea4d8ff      call    System.Numerics.Vector.ThrowInsufficientNumberOfElementsException(Int32) (00007ffc`4357ac18)
```

For some reason on the previous iteration JIT wasn't inlining the ctors properly and it was leading to significant stack spillage, but as of the latest iteration the codegen seems very good.

File: src/System.Private.CoreLib/shared/System/Numerics/Vector.cs [hunk 5]
```diff
@@ -716,23 +716,25 @@ private Vector(ref Register existingRegister)
...
         }
         
         /// <summary>
         /// Constructs a vector from the given <see cref="ReadOnlySpan{T}"/>. The span must contain at least <see cref="Vector{T}.Count"/> elements.
         /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public Vector(ReadOnlySpan<T> values)
             : this()
         {
-            ThrowHelper.ThrowForUnsupportedVectorBaseType<T>();
```
Comment (tannergooding) [inline src/System.Private.CoreLib/shared/System/Numerics/Vector.cs]: Why was this removed? This is a public method...

Comment (GrabYourPitchforks) [reply src/System.Private.CoreLib/shared/System/Numerics/Vector.cs]: Not needed in methods which access the `Count` property, as the property getter would throw. (This pattern is used elsewhere in this file.)

Comment (tannergooding) [reply src/System.Private.CoreLib/shared/System/Numerics/Vector.cs]: Thanks, I missed that bit.

**Comment by GrabYourPitchforks** [comment]
CI is reporting some unit test failures related to AVX2. These are unrelated to this PR and will be resolved by https://github.com/dotnet/coreclr/commit/54af92b34fc18ee9d8adada35d1f527c67224be7. I won't hold up merging this PR waiting for it.
